src/abc_atlas_access/abc_atlas_cache/cloud_cache.py

```diff
@@ -0,0 +1,1197 @@
+        """
+        Check if a file exists locally. If it does not, download it
+        and return True. Return False otherwise.
+
+        Parameters
+        ----------
+        file_attributes: CacheFileAttributes
+            Describes the file to download
+        force_download: bool
+            If True, force the file to be downloaded even if it already exists
+            locally
+        skip_hash_check: bool
+            If True, skip the file hash check. Not recommended as this
+            verifies that the file was downloaded successfully.
+
+        Returns
+        -------
+        bool
+            True if the file was downloaded; False otherwise
+
+        Raises
+        ------
+        RuntimeError
+            If the path to the directory where the file is to be saved
+            points to something that is not a directory.
+
+        RuntimeError
+            If it is not able to successfully download the file after
+            10 iterations
+        """
+        was_downloaded = False
+        if force_download and file_attributes.local_path.exists():
+            file_attributes.local_path.unlink()
+
+        local_path = file_attributes.local_path
+
+        local_dir = local_path.parents[0]
+        local_dir.mkdir(parents=True, exist_ok=True)
+        if not local_dir.is_dir():
+            raise RuntimeError(f"{local_dir}\n"
+                               "is not a directory")
+        obj_key = file_attributes.relative_path
+
+        n_iter = 0
+        max_iter = 10  # maximum number of times to try download
+
+        pbar = None
+        if not self._file_exists(file_attributes):
+            response = self.s3_client.list_objects_v2(Bucket=self._bucket_name,
+                                                      Prefix=str(obj_key))
+            object_info = response['Contents'][0]
+            pbar = tqdm.tqdm(desc=object_info["Key"].split("/")[-1],
+                             total=object_info["Size"],
+                             unit_scale=True,
+                             unit_divisor=1000.,
+                             unit="MB")
+
+        while not self._file_exists(file_attributes):
+            n_iter += 1
+            was_downloaded = True
+            response = self.s3_client.get_object(Bucket=self._bucket_name,
+                                                 Key=str(obj_key))
+
+            if 'Body' in response:
+                with open(local_path, 'wb') as out_file:
+                    for chunk in response['Body'].iter_chunks():
+                        out_file.write(chunk)
+                        pbar.update(len(chunk))
+
+            # Verify the hash of the downloaded file
+            test_checksum = file_hash_from_path(file_attributes.local_path)
+            if not skip_hash_check and \
+                    test_checksum != file_attributes.file_hash:
+                file_attributes.local_path.unlink(missing_ok=True)
+
+                if n_iter > max_iter:
```

Comment:
Do you mind putting this outside of `if not skip_hash_check`? I know that `n_iter` won't increment past 1 unless `skip_hash_check` is False. I think it is less confusing, though, if this is outside of the block.